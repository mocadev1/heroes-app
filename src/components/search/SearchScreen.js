import React, { useMemo } from 'react';
import queryString from 'query-string'

import HeroCard from '../heroes/HeroCard';
import useForm from '../../hooks/useForm';
import { useLocation } from 'react-router';
import { getHeroesByName } from '../../selectors/getHeroesByName';

const SearchScreen = ({ history }) => {

    const location = useLocation();

    const { q = '' } = queryString.parse( location.search )
    
    const [ { searchText }, handleInputChange ] = useForm({
        searchText: q
    });

    
    const filteredHeroes = useMemo( () =>  getHeroesByName( q ), [ q ] )


    
    const handleSearch = (e) => {
        e.preventDefault();
        history.push(`?q=${ searchText }`)
    }
    
    
    return (
        <div>
            <h1>Search Screen</h1>
            <hr />

            <div className="row">

                <div className="col-5">
                    <h4> Search Form </h4>
                    <hr />

                    <form
                        onSubmit={ handleSearch }
                        className="d-grid"
                    >

                        <input
                            type="text"
                            name="searchText"
                            placeholder="Find your hero!"
                            value={ searchText }
                            onChange={ handleInputChange }
                            autoComplete="off"
                            className="form-control"
                        />

                        <button
                            type="submit"
                            className="btn m-1 btn-block btn-outline-primary"
                        >
                            Search...
                        </button>
                        
                    </form>

                </div>

                <div className="col-7">

                    <h4> Results </h4>
                    <hr />

                    {
                        (q === '' )
                            &&
                            <div className="alert alert-info">
                                Search a Hero
                            </div>
                    }

                    {
                        (q !== '' && filteredHeroes.length === 0 )
                            &&
                            <div className="alert alert-danger">
                                There is no hero with { q }
                            </div>
                    }

                    {
                        filteredHeroes.map( hero => (
                            <HeroCard
                                key={ hero.id }
                                { ...hero}
                            />
                        )) 

                        
                    }
                    
                </div>


            </div>
        </div>
    )
}

export default SearchScreen