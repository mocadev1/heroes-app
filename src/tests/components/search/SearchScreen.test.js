import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router';
import SearchScreen from '../../../components/search/SearchScreen';

describe('<SearchScreen /> tests', () => {

    const historyMock = {
        push: jest.fn()
    }
    
    
    test('should render with the default values properly', () => {


        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route path="/search" component={ SearchScreen }/>
            </MemoryRouter>
        );

        

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('input').text().trim() ).toBe('');
        expect( wrapper.find('.alert-info').text().trim() ).toBe('Search a Hero');
        
    });
    
    test('should show Batman as result and the input should contain the queryString value', () => {


        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route path="/search" component={ SearchScreen }/>
            </MemoryRouter>
        );

        expect( wrapper.find('input').prop('value') ).toBe('batman');
        expect( wrapper.find('HeroCard').exists() ).toBe(true);
        
        
    });
    
    test('should show an alert containing the queryString value', () => {


        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=chapulin%20colorado']}>
                <Route path="/search" component={ SearchScreen }/>
            </MemoryRouter>
        );


        expect( wrapper.find('input').prop('value') ).toBe('chapulin colorado');
        expect( wrapper.find('HeroCard').exists() ).toBe(false);
        expect( wrapper.find('.alert-danger').text().trim() ).toBe(`There is no hero with chapulin colorado`);
        
        
    });

    test('should call the history PUSH method', () => {
        

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route path="/search" component={ () => <SearchScreen history={ historyMock } /> }/>
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change', {
            target: {
                name: 'searchText',
                value: 'spiderman'
            }
        });

        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        });

        expect( historyMock.push ).toHaveBeenCalledWith('?q=spiderman');


    });
    
    
    
    
});
