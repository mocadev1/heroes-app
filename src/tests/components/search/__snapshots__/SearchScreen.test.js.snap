// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchScreen /> tests should render with the default values properly 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
         Search Form 
      </h4>
      <hr />
      <form
        className="d-grid"
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero!"
          type="text"
          value=""
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
         Results 
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a Hero
      </div>
    </div>
  </div>
</div>
`;
